### PR TITLE
r/storagegateway_tape_pool - add new resource

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #10723

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_storagegateway_tape_pool: add new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSStorageGatewayTapePool_'
--- PASS: TestAccAWSStorageGatewayTapePool_basic (40.65s)
--- PASS: TestAccAWSStorageGatewayTapePool_retention (39.87s)
--- PASS: TestAccAWSStorageGatewayTapePool_tags (99.66s)
--- PASS: TestAccAWSStorageGatewayTapePool_disappears (29.18s)
```

### CODE REVIEW

File: aws/resource_aws_storagegateway_tape_pool.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 100),
+			},
+			"storage_class": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(storagegateway.TapeStorageClass_Values(), false),
+			},
+			"retention_look_type": {
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool.go:41]:
```suggestion
			"retention_lock_type": {
```

File: aws/resource_aws_storagegateway_tape_pool.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	}
+}
+
+func resourceAwsStorageGatewayTapePoolCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).storagegatewayconn
+
+	input := &storagegateway.CreateTapePoolInput{
+		PoolName:                aws.String(d.Get("pool_name").(string)),
+		StorageClass:            aws.String(d.Get("storage_class").(string)),
+		RetentionLockType:       aws.String(d.Get("retention_look_type").(string)),
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool.go:66]:
```suggestion
		RetentionLockType:       aws.String(d.Get("retention_lock_type").(string)),
```

File: aws/resource_aws_storagegateway_tape_pool.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		return nil
+	}
+
+	pool := output.PoolInfos[0]
+
+	poolArn := aws.StringValue(pool.PoolARN)
+	d.Set("arn", poolArn)
+	d.Set("pool_name", aws.StringValue(pool.PoolName))
+	d.Set("retention_lock_time_in_days", aws.Int64Value(pool.RetentionLockTimeInDays))
+	d.Set("retention_look_type", aws.StringValue(pool.RetentionLockType))
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool.go:122]:
```suggestion
	d.Set("retention_lock_type", aws.StringValue(pool.RetentionLockType))
```

File: aws/resource_aws_storagegateway_tape_pool_test.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+		CheckDestroy: testAccCheckAWSStorageGatewayTapePoolDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSStorageGatewayTapePoolBasicConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSStorageGatewayTapePoolExists(resourceName, &TapePool),
+					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "storagegateway", regexp.MustCompile(`tapepool/pool-.+`)),
+					resource.TestCheckResourceAttr(resourceName, "pool_name", rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_class", "GLACIER"),
+					resource.TestCheckResourceAttr(resourceName, "retention_look_type", "NONE"),
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool_test.go:32]:
```suggestion
					resource.TestCheckResourceAttr(resourceName, "retention_lock_type", "NONE"),
```

File: aws/resource_aws_storagegateway_tape_pool_test.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+		CheckDestroy: testAccCheckAWSStorageGatewayTapePoolDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSStorageGatewayTapePoolRetentionConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSStorageGatewayTapePoolExists(resourceName, &TapePool),
+					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "storagegateway", regexp.MustCompile(`tapepool/pool-.+`)),
+					resource.TestCheckResourceAttr(resourceName, "pool_name", rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_class", "GLACIER"),
+					resource.TestCheckResourceAttr(resourceName, "retention_look_type", "GOVERNANCE"),
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool_test.go:62]:
```suggestion
					resource.TestCheckResourceAttr(resourceName, "retention_lock_type", "GOVERNANCE"),
```

File: aws/resource_aws_storagegateway_tape_pool_test.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+}
+`, rName)
+}
+
+func testAccAWSStorageGatewayTapePoolRetentionConfig(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_storagegateway_tape_pool" "test" {
+  pool_name                   = %[1]q
+  storage_class               = "GLACIER"
+  retention_look_type         = "GOVERNANCE"
```
Comment (ewbankkit) [inline aws/resource_aws_storagegateway_tape_pool_test.go:210]:
```suggestion
  retention_lock_type         = "GOVERNANCE"
```

File: website/docs/r/storagegateway_tape_pool.html.markdown [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `pool_name` - (Required) The name of the new custom tape pool.
+* `storage_class` - (Required) The storage class that is associated with the new custom pool. When you use your backup application to eject the tape, the tape is archived directly into the storage class that corresponds to the pool. Possible values are `DEEP_ARCHIVE` or `GLACIER`.
+* `retention_look_type` - (Required) Tape retention lock can be configured in two modes. When configured in governance mode, AWS accounts with specific IAM permissions are authorized to remove the tape retention lock from archived virtual tapes. When configured in compliance mode, the tape retention lock cannot be removed by any user, including the root AWS account. Possible values are `COMPLIANCE`, `GOVERNANCE`, and `NONE`. Default value is `NONE`.
```
Comment (ewbankkit) [inline website/docs/r/storagegateway_tape_pool.html.markdown:28]:
```suggestion
* `retention_lock_type` - (Required) Tape retention lock can be configured in two modes. When configured in governance mode, AWS accounts with specific IAM permissions are authorized to remove the tape retention lock from archived virtual tapes. When configured in compliance mode, the tape retention lock cannot be removed by any user, including the root AWS account. Possible values are `COMPLIANCE`, `GOVERNANCE`, and `NONE`. Default value is `NONE`.
```

**Comment by DrFaust92** [comment]
Thanks @ewbankkit, also fixed some post test delete issues.

```
--- PASS: TestAccAWSStorageGatewayTapePool_disappears (34.21s)
--- PASS: TestAccAWSStorageGatewayTapePool_retention (48.40s)
--- PASS: TestAccAWSStorageGatewayTapePool_basic (48.84s)
--- PASS: TestAccAWSStorageGatewayTapePool_tags (123.17s)
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSStorageGatewayTapePool_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSStorageGatewayTapePool_ -timeout 120m
=== RUN   TestAccAWSStorageGatewayTapePool_basic
=== PAUSE TestAccAWSStorageGatewayTapePool_basic
=== RUN   TestAccAWSStorageGatewayTapePool_retention
=== PAUSE TestAccAWSStorageGatewayTapePool_retention
=== RUN   TestAccAWSStorageGatewayTapePool_tags
=== PAUSE TestAccAWSStorageGatewayTapePool_tags
=== RUN   TestAccAWSStorageGatewayTapePool_disappears
=== PAUSE TestAccAWSStorageGatewayTapePool_disappears
=== CONT  TestAccAWSStorageGatewayTapePool_basic
=== CONT  TestAccAWSStorageGatewayTapePool_disappears
=== CONT  TestAccAWSStorageGatewayTapePool_tags
=== CONT  TestAccAWSStorageGatewayTapePool_retention
--- PASS: TestAccAWSStorageGatewayTapePool_disappears (18.51s)
--- PASS: TestAccAWSStorageGatewayTapePool_basic (26.50s)
--- PASS: TestAccAWSStorageGatewayTapePool_retention (27.10s)
--- PASS: TestAccAWSStorageGatewayTapePool_tags (55.03s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	55.084s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TESTARGS='-run=TestAccAWSStorageGatewayTapePool_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSStorageGatewayTapePool_ -timeout 120m
=== RUN   TestAccAWSStorageGatewayTapePool_basic
=== PAUSE TestAccAWSStorageGatewayTapePool_basic
=== RUN   TestAccAWSStorageGatewayTapePool_retention
=== PAUSE TestAccAWSStorageGatewayTapePool_retention
=== RUN   TestAccAWSStorageGatewayTapePool_tags
=== PAUSE TestAccAWSStorageGatewayTapePool_tags
=== RUN   TestAccAWSStorageGatewayTapePool_disappears
=== PAUSE TestAccAWSStorageGatewayTapePool_disappears
=== CONT  TestAccAWSStorageGatewayTapePool_basic
=== CONT  TestAccAWSStorageGatewayTapePool_tags
=== CONT  TestAccAWSStorageGatewayTapePool_retention
=== CONT  TestAccAWSStorageGatewayTapePool_disappears
--- PASS: TestAccAWSStorageGatewayTapePool_disappears (11.63s)
--- PASS: TestAccAWSStorageGatewayTapePool_retention (16.05s)
--- PASS: TestAccAWSStorageGatewayTapePool_basic (16.31s)
--- PASS: TestAccAWSStorageGatewayTapePool_tags (38.87s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	40.314s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TESTARGS='-run=TestAccAWSStorageGatewayTapePool_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSStorageGatewayTapePool_ -timeout 120m
=== RUN   TestAccAWSStorageGatewayTapePool_basic
=== PAUSE TestAccAWSStorageGatewayTapePool_basic
=== RUN   TestAccAWSStorageGatewayTapePool_retention
=== PAUSE TestAccAWSStorageGatewayTapePool_retention
=== RUN   TestAccAWSStorageGatewayTapePool_tags
=== PAUSE TestAccAWSStorageGatewayTapePool_tags
=== RUN   TestAccAWSStorageGatewayTapePool_disappears
=== PAUSE TestAccAWSStorageGatewayTapePool_disappears
=== CONT  TestAccAWSStorageGatewayTapePool_basic
=== CONT  TestAccAWSStorageGatewayTapePool_disappears
=== CONT  TestAccAWSStorageGatewayTapePool_tags
=== CONT  TestAccAWSStorageGatewayTapePool_retention
--- PASS: TestAccAWSStorageGatewayTapePool_disappears (8.66s)
2020/10/29 10:51:45 [DEBUG] Trying to get account information via sts:GetCallerIdentity
--- PASS: TestAccAWSStorageGatewayTapePool_retention (12.12s)
--- PASS: TestAccAWSStorageGatewayTapePool_basic (12.67s)
--- PASS: TestAccAWSStorageGatewayTapePool_tags (28.68s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	30.118s
```

**Pull request merged**